### PR TITLE
HID: sony: Enable Bluetooth Gasia third-party PS3 controllers

### PR DESCRIPTION
**Pull request by vs7**

### CODE REVIEW

File: drivers/hid/hid-sony.c [hunk 1]
```diff
@@ -1067,8 +1067,24 @@ static int sixaxis_set_operational_usb(struct hid_device *hdev)
 static int sixaxis_set_operational_bt(struct hid_device *hdev)
 {
 	unsigned char buf[] = { 0xf4,  0x42, 0x03, 0x00, 0x00 };
-	return hid_hw_raw_request(hdev, buf[0], buf, sizeof(buf),
-				  HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+	unsigned char ack;
+	ack = hid_hw_raw_request(hdev, buf[0], buf, sizeof(buf), HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
```
Comment (dlech) [inline drivers/hid/hid-sony.c]: ~~This `ack` value is ignored.~~

File: drivers/hid/hid-sony.c [hunk 2]
```diff
@@ -1067,8 +1067,24 @@ static int sixaxis_set_operational_usb(struct hid_device *hdev)
...
+	unsigned char enable_gasia[] = {
```
Comment (dlech) [inline drivers/hid/hid-sony.c:1071]: Declaring a variable here will cause a compiler warning. It needs to be declared at the beginning with the rest of the variables.

File: drivers/hid/hid-sony.c [hunk 3]
```diff
@@ -1067,8 +1067,24 @@ static int sixaxis_set_operational_usb(struct hid_device *hdev)
...
+			0xA2,
+			0x01,
+			0x00, 0x00, 0x00, 0x00, 0x00,   // rumble values [0x00, right-timeout, right-force, left-timeout, left-force]
```
Comment (dlech) [inline drivers/hid/hid-sony.c]: code style: Linux kernel does not use C99 comments (`//`). Please use `/* ... */`.

Also, keep lines <= 80 chars if possible.

File: drivers/hid/hid-sony.c [hunk 4]
```diff
@@ -1067,8 +1067,24 @@ static int sixaxis_set_operational_usb(struct hid_device *hdev)
...
+			0xff, 0x27, 0x10, 0x00, 0x32,   // LED 4
+			0xff, 0x27, 0x10, 0x00, 0x32,   // LED 3
+			0xff, 0x27, 0x10, 0x00, 0x32,   // LED 2
+			0xff, 0x27, 0x10, 0x00, 0x32,   // LED 1
+			0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00
+	};
+	hid_hw_raw_request(hdev, enable_gasia[0], enable_gasia, sizeof(enable_gasia), HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
```
Comment (dlech) [inline drivers/hid/hid-sony.c:1088]: What happens if we send this request to a controller that does not support it? Do we get a bad `ack` value?

File: drivers/hid/hid-sony.c (hunk 1 above)
Comment (dlech) [inline drivers/hid/hid-sony.c]: Never mind, I see you are not using it again below.

File: drivers/hid/hid-sony.c (hunk 4 above)
Comment (dlech) [inline drivers/hid/hid-sony.c:1088]: I guess this doesn't matter since you don't have `ack =` here.

Comment (vs7) [inline drivers/hid/hid-sony.c:1088]: This hack is based on https://github.com/supertypo/qtsixa/commit/bbcc5c36d1f317509c39e9e84e77599fea4d81ed
I've only tested it with Gasia (fake PS3) joystick.
`ack` needs to be returned for original dualshock (see `sony_probe`)

**Comment by dlech** [comment]
Looks good to me. Please fix the variable declarations and comments and I will merge.

**Comment by dlech** [comment]
Here is an upstream patch that seems to do the same thing:

https://github.com/torvalds/linux/commit/a7de9b867be0662e72db69fd40a4f1d2aa80ecc0

Does this get your controller working instead of this pull request?

**Comment by vs7** [comment]
That patch deals only with USB connection, not BT. I've checked that - it works, when cable plugged,

**Comment by dlech** [comment]
Thanks!

**Pull request merged**